app/services/network_topology_service.rb

```diff
@@ -0,0 +1,117 @@
+class NetworkTopologyService < TopologyService
+  def initialize(provider_id)
+    @provider_id = provider_id
+    @providers = retrieve_providers(ManageIQ::Providers::NetworkManager, @provider_id)
+    # @entity = retrieve_entity(entity_id, entity_type)
+    @entity = @providers.first
+  end
+
+  def entity_type(entity)
+    if entity.kind_of?(CloudNetwork)
+      entity.class.base_class.name.demodulize
+    else
+      entity.class.name.demodulize
+    end
+  end
+
+  def build_topology
+    topo_items = {}
+    links = []
+
+    entity_relationships = {
+      :NetworkManager => {
+        :CloudSubnets => {
+          :CloudNetwork => nil,
+          :Vms => {
+            :FloatingIps => nil,
+            :CloudTenant => nil,
+            :SecurityGroups => nil
+          },
+          :NetworkRouter => {
+            :CloudNetwork => nil
+          },
+        }
+      },
+    }
+    # loaded = @providers.includes(:cloud_subnets => [:cloud_network,
```

Comment:
@abonas, cc @Fryguy so as a start, you can just do this to get rid of the N+1 (actually this is (N+1) ^x, where x is height of the tree, so pretty bad)

I assume the tree builder does something similar?

Btw. I can;t use it now because of my favorite rails bug  https://github.com/rails/rails/pull/24192

@Fryguy also seems like there is something weird with HABTM, cause I still have O(N) queries for habtm relation, it's the Vm has_and_belongs_to_many security_groups through network_port. Check  http://paste.openstack.org/show/492717/

@abonas btw. to see what queries your code does, when running from console, you can use this code ActiveRecord::Base.logger = Logger.new(STDOUT)